bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/ResultRowAdapter.java

```diff
@@ -0,0 +1,181 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.hbase.adapters.read;
+
+import com.google.api.core.InternalApi;
+import com.google.cloud.bigtable.data.v2.models.RowAdapter;
+import com.google.cloud.bigtable.hbase.util.ByteStringer;
+import com.google.cloud.bigtable.hbase.util.TimestampConverter;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.ByteString;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.client.Result;
+
+/**
+ * Adapter for {@link RowAdapter} that uses {@link Result} to represent logical rows.
+ *
+ * <p>For internal use only - public for technical reasons.
+ */
+@InternalApi("For internal usage only")
+public class ResultRowAdapter implements RowAdapter<Result> {
+
+  @Override
+  public RowBuilder<Result> createRowBuilder() {
+    return new ResultRowBuilder();
+  }
+
+  @Override
+  public boolean isScanMarkerRow(Result result) {
+    return result.isEmpty();
+  }
+
+  @Override
+  public ByteString getKey(Result result) {
+    return ByteStringer.wrap(result.getRow());
+  }
+
+  public static class ResultRowBuilder implements RowBuilder<Result> {
+    private byte[] currentKey;
+    private String family;
+    private ByteString qualifier;
+    private List<String> labels;
+    private long timestamp;
+    private ByteString value;
+
+    private Map<String, List<RowCell>> cells = new TreeMap<>();
+    private List<RowCell> currentFamilyCells = null;
+    private String previousFamily;
+    private int totalCellCount = 0;
+
+    @Override
+    public void startRow(ByteString rowKey) {
+      this.currentKey = ByteStringer.extract(rowKey);
+    }
+
+    @Override
+    public void startCell(
+        String family, ByteString qualifier, long timestamp, List<String> labels, long size) {
+      this.family = family;
+      this.qualifier = qualifier;
+      this.timestamp = timestamp;
+      this.labels = labels;
+      this.value = ByteString.EMPTY;
+    }
+
+    @Override
+    public void cellValue(ByteString value) {
+      this.value = this.value.concat(value);
+    }
+
+    /**
+     * Adds a Cell to {@link RowCell}'s map which is ordered by family. cells received from {@link
+     * RowBuilder} has ordering as:
+     *
+     * <ul>
+     *   <li>family names clustered, but not sorted
+     *   <li>qualifiers in each family cluster is sorted lexicographically
+     *   <li>then descending by timestamp
+     * </ul>
+     *
```

Comment:
you need to rephrase the order, it's not clear at all.
Probably you meant: Cells with the the same family are sorted according to:
- by clustered/not clustered
- by qualifiers (lexicographically)
- by timestamp (descending)

perhaps you can skip this paragraph, you have the explanation further